clippy_lints/src/implicit_transmute_types.rs

```diff
@@ -0,0 +1,105 @@
+use rustc_errors::Applicability;
+use rustc_hir::{self as hir, ExprKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for calls to [`transmute`] without explicit type parameters
+    /// (i.e. without turbofish syntax).
+    ///
+    /// ### Why is this bad?
+    /// In most cases Rust's type inference is helpful, however it can cause
+    /// problems with [`transmute`]. `transmute` is wildly unsafe
+    /// unless the types being transmuted are known to be compatible. As such,
+    /// a seemingly innocent change in something's type can end up making a
+    /// previously-valid transmute suddenly become unsound. Thus it is
+    /// good practice to always be explicit about the types you expect to be
+    /// transmuting between, so that the compiler will force you to
+    /// reexamine the transmute if either type changes.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[repr(transparent)]
+    /// struct CharWrapper {
+    ///     _inner: char,
+    /// }
+    ///
+    /// let wrapped = CharWrapper { _inner: 'a' };
+    /// let transmuted = unsafe { core::mem::transmute(wrapped) };
+    ///
+    /// // This is sound now, but if it gets changed in the future to
+    /// // something that expects a type other than `char`, the transmute
+    /// // would infer it returns that type, which is likely unsound.
+    /// let _ = char::is_lowercase(transmuted);
+    /// ```
+    ///
+    /// Specify type parameters:
+    /// ```rust
+    /// # #[repr(transparent)]
+    /// # struct CharWrapper {
+    /// #     _inner: char,
+    /// # }
+    /// let wrapped = CharWrapper { _inner: 'a' };
+    /// // Because we explicitly specify the types for the transmute, any change in
+    /// // surrounding code that would cause the transmute call to infer different
+    /// // types will now be caught by typechecking, forcing us to come back and
+    /// // reassess the soundsness of transmuting between the new types.
+    /// let transmuted = unsafe { core::mem::transmute::<CharWrapper, char>(wrapped) };
+    ///
+    /// let _ = char::is_lowercase(transmuted);
+    /// ```
+    ///
+    /// If you decide that you *do* want the types to be inferred,
+    /// you can silence the lint by conveying your intention explicitly:
```

Comment:
I feel like having a configuration option to also disallow this would be nice, you can probably check the 1st and 2nd arguments of the last segment to see if either are `_`